Packet++/src/SomeIpLayer.cpp

```diff
@@ -0,0 +1,379 @@
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->sessionID = htobe16(sessionID);
+}
+
+uint16_t SomeIpLayer::getClientID() const
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	return be16toh(hdr->clientID);
+}
+
+void SomeIpLayer::setClientID(uint16_t clientID)
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->clientID = htobe16(clientID);
+}
+
+uint8_t SomeIpLayer::getProtocolVersion() const
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	return hdr->protocolVersion;
+}
+
+void SomeIpLayer::setProtocolVersion(uint8_t version)
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->protocolVersion = version;
+}
+
+uint8_t SomeIpLayer::getInterfaceVersion() const
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	return hdr->interfaceVersion;
+}
+
+void SomeIpLayer::setInterfaceVersion(uint8_t version)
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->interfaceVersion = version;
+}
+
+uint8_t SomeIpLayer::getMessageType() const
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	return hdr->msgType;
+}
+
+void SomeIpLayer::setMessageType(MsgType type)
+{
+	setMessageType((uint8_t)type);
+}
+
+void SomeIpLayer::setMessageType(uint8_t type)
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->msgType = type;
+}
+
+uint8_t SomeIpLayer::getReturnCode() const
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	return hdr->returnCode;
+}
+
+void SomeIpLayer::setReturnCode(uint8_t returnCode)
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->returnCode = returnCode;
+}
+
+std::string SomeIpLayer::toString() const
+{
+	std::stringstream dataStream;
+	dataStream << std::hex;
+
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	dataStream << "-- SomeIp Header --" << std::endl;
+	dataStream << "Service ID \t\t 0x" << be16toh(hdr->serviceID) << std::endl
+			   << "Method ID \t\t 0x" << be16toh(hdr->methodID) << std::endl
+			   << "Length \t\t\t 0x" << be32toh(hdr->length) << std::dec << " (" << be32toh(hdr->length) << " bytes)"
+			   << std::hex << std::endl
+			   << "Client ID \t\t 0x" << be16toh(hdr->clientID) << std::endl
+			   << "Session ID \t\t 0x" << be16toh(hdr->sessionID) << std::endl
+			   << "Protocol Version \t 0x" << static_cast<int>(hdr->protocolVersion) << std::endl
+			   << "Interface Version \t 0x" << static_cast<int>(hdr->interfaceVersion) << std::endl
+			   << "Message Type \t\t 0x" << static_cast<int>(hdr->msgType) << std::endl
+			   << "Return Code \t\t 0x" << static_cast<int>(hdr->returnCode) << std::endl;
+
+	return dataStream.str();
+}
+
+void SomeIpLayer::parseNextLayer()
+{
+	size_t headerLen = getHeaderLen();
+	if (m_DataLen <= headerLen)
+		return;
+
+	uint8_t *payload = m_Data + headerLen;
+	size_t payloadLen = m_DataLen - headerLen;
+
+	// TODO: handle SomeIPSD/SomeIPTP packets
+
+	if (isSomeIpProtocol(payload, payloadLen))
+	/* else parse as SomeIpLayer (i.e. respect chaining of SomeIpLayers in one Ethernet frame) */
+	{
+		m_NextLayer = new SomeIpLayer(payload, payloadLen, this, m_Packet);

```

Comment:
Just for my understanding. There are some comments related to the SOME/IP service discovery. But here we assume that next layer is also plain SOME/IP layer. I was thinking about making a separate layer for SOME/IP SD messages. Please excuse me if I misunderstood the logic.

Reply:
We currently removed the code which creates the SOME/IP-SD layer, because we first wanted to upstream the base protocol, and later extend it with the SD and TP implementation. We decided to not upstream commented out code, which is why the reference to SD and TP is missing here. Sorry for the confusion.
The question why we are parsing further payload as a plain SOME/IP layer is a very good one. AUTOSAR specifies that a UDP datagram can contain multiple SOME/IP layers behind each other. This code implements this logic for the plain SOME/IP protocol.

Reply:
Frankly, on practice never seen nested SOME/IP packets except SD messages (which is a bit of different stuff). Yes, you are right, I have forgotten about this 

https://www.autosar.org/fileadmin/user_upload/standards/foundation/1-0/AUTOSAR_PRS_SOMEIPProtocol.pdf
```
[PRS_SOMEIP_00140] d The header format allows transporting more than one
SOME/IP message in a single UDP packet. The SOME/IP implementation shall identify
the end of a SOME/IP message by means of the SOME/IP length field. Based on the
UDP length field, SOME/IP shall determine if there are additional SOME/IP messages
in the UDP packet. c(RS_SOMEIP_00010)
```